packages/datadog-plugin-gcloud-pubsub/test/index.spec.js

```diff
@@ -0,0 +1,153 @@
+        })
+
+        describe('createTopic', () => {
+          it('should be instrumented', () =>
+            performTest({
+              meta: { 'pubsub.method': 'createTopic' }
+            }, async ({ pubsub, topicName }) => {
+              await pubsub.createTopic(topicName)
+            })
+          )
+        })
+
+        describe('publishMessage', () => {
+          it('should be instrumented', () =>
+            performTest({
+              meta: { 'pubsub.method': 'createTopic' }
+            }, async ({ pubsub, topicName }) => {
+              const [topic] = await pubsub.createTopic(topicName)
+              return topic.publishMessage({ data: Buffer.from('hello') })
+            })
+          )
+        })
+
+        describe('onMessage', () => {
+          it('should be instrumented', () =>
+            performTest({
+              name: 'gpubsub.onmessage'
+            }, async ({ pubsub, topicName }) => {
+              const [topic] = await pubsub.createTopic(topicName)
+              const [sub] = await topic.createSubscription('foo')
+              sub.on('message', () => {})
+              return topic.publishMessage({ data: Buffer.from('hello') })
+            })
+          )
+
+          it('should give the current span a parentId from the sender', () =>
+            performTest({
+              name: 'gpubsub.onmessage'
+            }, async ({ pubsub, topicName }) => {
+              const [topic] = await pubsub.createTopic(topicName)
+              const [sub] = await topic.createSubscription('foo')
+              const rxPromise = new Promise((resolve, reject) => {
+                sub.on('message', () => {
+                  const receiverSpanContext = tracer.scope().active()._spanContext
+                  try {
+                    expect(receiverSpanContext._parentId).to.be.an('object')
+                    resolve()
+                  } catch (e) {
+                    reject(e)
+                  }
+                })
+              })
+              await topic.publishMessage({ data: Buffer.from('hello') })
+              return rxPromise
+            })
+          )
+        })
+      })
+
+      describe('with configuration', () => {
+        beforeEach(() => {
+          tracer = require('../../dd-trace')
+          agent.load(plugin, '@google-cloud/pubsub', {
+            service: 'a_test_service'
+          })
+          gpubsub = require(`../../../versions/@google-cloud/pubsub@${version}`).get()
+        })
+
+        describe('createTopic', () => {
+          it('should be instrumented', () =>
+            performTest({
+              service: 'a_test_service',
+              meta: { 'pubsub.method': 'createTopic' }
+            }, async ({ pubsub, topicName }) => {
+              await pubsub.createTopic(topicName)
+            })
+          )
+        })
+      })
+    })
+  })
+
+  function performTest (expected, test) {
```

Comment:
Not a blocker, but this helper makes the tests a bit difficult to read. The flow of each test is reversed, and it's hard to see what is what, and what the test is actually testing. While shorter tests are a good thing, I feel this ends up hiding what is tested. This can always be improved later but worth discussing if there is a quick win here.

Reply:
Here's a proposed refactor: I'll move the creation of the pubsub object into a `beforeEach`, and rework `performTest` so that all it does is the expectation in an `agent.use`, with defaults, and rename it to `expectOneSpan` like in the `fs` tests.